File: billing-application/src/main/java/com/example/billing/application/command/SettlementResult.java
package com.example.billing.application.command;

import com.example.billing.domain.settlement.BillingPeriod;
import com.example.billing.domain.shared.CustomerId;

import java.util.UUID;

public record SettlementResult(
        CustomerId customerId,
        BillingPeriod period,
        UUID invoiceId,
        boolean alreadyProcessed,
        boolean paymentSucceeded,
        String message
) {
    public static SettlementResult skipped(CustomerId customerId, BillingPeriod period, String reason) {
        return new SettlementResult(customerId, period, null, true, false, reason);
    }
    public static SettlementResult success(CustomerId customerId, BillingPeriod period,
                                           UUID invoiceId, boolean paymentSucceeded) {
        return new SettlementResult(customerId, period, invoiceId, false, paymentSucceeded,
                paymentSucceeded ? "paid" : "invoice issued, payment failed");
    }
}

File: billing-application/src/main/kotlin/com/example/billing/application/command/SettlementResult.kt
package com.example.billing.application.command

import com.example.billing.domain.settlement.BillingPeriod
import com.example.billing.domain.shared.CustomerId
import java.util.UUID

@JvmRecord
data class SettlementResult(
    val customerId: CustomerId,
    val period: BillingPeriod,
    val invoiceId: UUID?,
    val alreadyProcessed: Boolean,
    val paymentSucceeded: Boolean,
    val message: String,
) {
    companion object {
        @JvmStatic
        fun skipped(customerId: CustomerId, period: BillingPeriod, reason: String): SettlementResult =
            SettlementResult(customerId, period, null, true, false, reason)

        @JvmStatic
        fun success(
            customerId: CustomerId,
            period: BillingPeriod,
            invoiceId: UUID,
            paymentSucceeded: Boolean,
        ): SettlementResult = SettlementResult(
            customerId, period, invoiceId, false, paymentSucceeded,
            if (paymentSucceeded) "paid" else "invoice issued, payment failed",
        )
    }
}
